main/main.c

```diff
@@ -320,8 +320,35 @@ static PHP_INI_MH(OnChangeMemoryLimit)
 	if (new_value) {
 		value = zend_ini_parse_uquantity_warn(new_value, entry->name);
 	} else {
-		value = Z_L(1)<<30;		/* effectively, no limit */
+		value = Z_L(1) << 30; /* effectively, no limit */
 	}
+
+	/* If max_memory_limit is not set to unlimited, verify change */
+	if (PG(max_memory_limit) != -1) {
+		if (value == -1) {
+			zend_error(
+				stage == ZEND_INI_STAGE_STARTUP ? E_ERROR : E_WARNING,
```

Comment:
Looking at this again, I think it would be better to stick to a warning in all cases. Erroring is very harsh, especially on startup, taking down the entire website.

Reply:
I'll change all cases to a warning :)